### ISSUE TITLE
catalog specs redux

### ISSUE THREAD

**Issue by jgraettinger** [estuary/flow#68]
Tracking issue for planned changes to catalog specifications:

  * [x] `materializationTarget` will be reworked to `endpoint`, which is meaningful both for captures and materializations. It has a *name* and *configuration* (parameters specific to the driver).
  * [x] `materializations` will be a top-level entity, and binds a *target* name, *collection*, and *endpoint* with *configuration*. The (target, collection, endpoint) tuple must be unique. Target is intended to be meaningful to the endpoint (eg, a topic or table name). 
  * [x] `captures` will be a top-level entity, and binds a *source* name, *collection*, and *endpoint* with *configuration* (like materializations). The (source, collection, endpoint) tuple must be unique. Source is meaningful to the endpoint (eg, a source table for CDC).
  * [x] Allow cycles in derivation dependencies (allow both A -> B -> A and A -> A) during catalog builds.
  * [x] Change representation of collections in flow.yaml to be an object instead of an array. The keys will be the collection name.
  * [x] Allow cyclic imports between catalog resources (implied requirement of derivation cycles).
  * [x] Verify referential integrity at the _end_ of build stages, so we can present a list of errors rather than fail immediately with just one (c.f. [pragma integrity_check](https://www.sqlite.org/pragma.html#pragma_integrity_check) and [foreign_key_check](https://www.sqlite.org/pragma.html#pragma_foreign_key_check))
  * [x] Similarly, lift current fail-fast triggers to inspection passes which collect and report multiple errors.
  * [x] Eliminate internal DB identifiers (ex resource_id, collection_id). Instead, use content-addressed primary keys (collection_name, canonical resource uri). Goal is to make database tables an assemblage of pure "facts" learned from specs, which can be independently manipulated during processing without regard to cross-table DB states (like rowids).
  * [x] Similarly, directly embed JSON specs which don't _need_ to be bespoke tables (lambdas, bootstraps, partition selectors, etc). Only named resources should be table entities.
  * [x] Enable modular SQL test (ideally using insta, and without using sqlite CLI).
  * [x] Separate current `catalog` crate into `catalog-models` (concerned with representation only), vs `catalog-build` (concerned with build passes).

Guiding principles:
 * The `flowctl materialize` workflow is a step backwards from our core design goal of keeping everything declarative. The help it offers (eg, enumerating projections) can be accomplished with other tools (eg, perhaps a `flowctl show projections`). Better to require the user to be explicit about materializations that exist and their projections, and to take verification / initialization actions during a future `flowctl deploy` workflow.
 * Being able to easily "round-trip" data from another system (that may be considered source-of-truth), through Flow transforms, and right back into that system is core to the Flow value prop. Specs should embrace a framing where endpoints are both producers and consumers.

**Comment by psFried** [comment]
I really like the symmetry with this design. 

I'm starting to think we should also define explicit endpoints for the HTTP and websocket endpoints in flow-ingester (i.e. always required the REST/websocket endpoints to be named in `captures` if you want to use them). If we're going to have a user list out other types of endpoints explicitly (e.g. s3 buckets, kafka topics), it would seem weird to _not_ list the REST and websocket endpoints, too. And I think if I were responsible for a production instance, I wouldn't want any "extra" ingestion methods enabled. What do you think?

**Comment by jgraettinger** [comment]
I think that’s a great idea. If you want API ingestion, you should declare
it like any other means.

**Comment by jgraettinger** [comment]
One expanded note on `materialization` -- the natural thing would be to have a materialization take a `source` spec as we currently define it -- having a name, optional schema, and partitions. I think that's right. One _should_ be able to materialize only a partitioned subset of the collection, or to apply a read-time schema that's doing more restrictive validations. Things that would otherwise require another derive-and-filter pass.

**Comment by psFried** [comment]
There's some other topics I'd like to discuss related to this. At this point, it's somewhat an
exercise in looking out at the horizon, but I think it could have implications on the here and now.

I was thinking we should plan on some `endpoint` types needing a JSON object in order to represent all
their connection information. Thinking about something like S3, which may need additional
information like credentials or server-side encryption keys, which we may not be able to include in
the URI. Do you think this is too speculative?

This also relates to the general problem of "how should we handle credentials", and I'd like some
feedback on this as well. I think it makes sense to just have credentials included directly in the
YAML and stored (for now) unencrypted in the catalog. The idea being that users could just use
`git-secret` or something similar to manage the yaml that has their sensitive credentials in a YAML
file that imports their "main" specs and just includes their `endpoints` definitions. For example:

`production.flow.yaml`:
```yaml
import:
  - base.flow.yaml
endpoints:
  thePostgresDB: {postgres: "postgres://username:superSecretPassword@foo.test:5432"}
```

A potential problem this creates is what happens when `base.flow.yaml` includes a materailization
that references `thePostgresDB` _before_ the production credentials have been specified in
`production.flow.yaml`. I think it makes sense to sidestep this problem entirely by always
registering all `endpoints` before processing any `imports`. The benefit being that we could still
have the catalog schema enforce the foreign key constraints for `materializations` and `captures`
referencing `endpoints`, while also enforcing that each `endpoint` is defined only once. 

An alternative is to require `base.flow.yaml` to _also_ define a `thePostgresDB` endpoint, and to
have the one from `production.flow.yaml` overwrite it (or somehow take precedence). But I think that
would open up a lot of potential for confusion and mistakes. Having endpoints be "injected" provides
a usable pattern for handling different environments, and makes it more difficult to use the wrong
endpoint accidentally since you'd be using something like `--source <production|test|whatever>.flow.yaml`
on the CLI.

Going back to the topic of encryption of credentials in the catalog, I have a sense that this might
be something we'll want to do _eventually_. I think it's fine to leave them unencrypted for now. If
you have access to the yaml that produces a catalog, then you already have the credentials. So it's
really just about communicateing to users that their catalog is just as sensitive as the yaml that
produced it, which ought to be a pretty straight forward concept.

**Comment by jgraettinger** [comment]
A couple of comments. First, the question of configuring endpoints is pretty well-trod ground in gazette. Check out: https://github.com/gazette/core/blob/master/broker/fragment/store_s3.go#L23-L44

tl;dr is that query parameters work well for configuration, especially when they only need to _name_ credentials provided via application default credentials. In the limited sense we should certainly keep compatibility with how Gazette thinks about credentials for buckets, etc.

More broadly, the pattern of having query parameters name a credential configuration that's given via a separate secret, bound at runtime is a pretty good one (eg as happens with application default credentials in each of the AWS vs GCP cloud flavors) ?

Regarding import vs endpoints ordering, agreed, we process endpoints, then imports, then collections, then captures & materializations.

**Comment by jgraettinger** [comment]
>  I think it makes sense to just have credentials included directly in the
YAML and stored (for now) unencrypted in the catalog.

Also makes sense to me. I think we'll live in a world where cloud application credentials are managed separately from eg postgres credentials for quite a while. I haven't thought deeply about it, but we may _never_ leave that world (as users can instead allow access to _their_ bucket to _our_ service account / IAM account, vs our requiring bespoke credentials from them)

**Comment by psFried** [comment]
> tl;dr is that query parameters work well for configuration

Ok this is good, and I think means that we can KISS for now and just make everything a URI until we have a solid use case for something else. In any case, I think we'll discover some new constraints/requirements once we start designing a multi-tenant SAAS, which for me is a little too far out to design for at this point.

**Comment by psFried** [comment]
When I first read this:

> One expanded note on materialization -- the natural thing would be to have a materialization take a source spec as we currently define it -- having a name, optional schema, and partitions. I think that's right. One should be able to materialize only a partitioned subset of the collection, or to apply a read-time schema that's doing more restrictive validations. Things that would otherwise require another derive-and-filter pass.

It seemed like such an obviously good idea that I didn't even bother to respond. But now that I've gotten a little closer, I've realized that I don't think we can allow materializations to do any sort of filtering. I'm having trouble explaining the issue succinctly, so I'll just provide an example.

```yaml
collection:
  - name: foo
    schema:
      type: object
      properties:
        id: { type: integer }
        state: { enum: [A, B] }
        sumOfSomething: { type: integer, reduce: { strategy: sum } }
      required: [id, state]
      reduce: { strategy: merge }
    key: [/id]
    projections:
      state:
        location: /state
        partition: true
...
materializations:
  - source:
      name: foo
      partitions:
        include:
          state: [A]
    endpoint: ourPostgresDB
    target: example_table
```

We'd expect that the table in postgres would only include documents that have `state: A`, but the
problem is that the `state` for a given id may switch between A and B. If we don't store all the
documents with a `state` of B, then it wouldn't be practical to compute the `sumOfSomething`
for a document that was in `state: B` and then switched to A. You'd either need to keep the current
state in some _other_ place, or you'd need to re-process all the documents with that `id`.

I'm sure we could figure out some way to make it work, maybe by using registers for the full
flow document. Or maybe by requiring that all partitions are materialized _somewhere_, sort of like
how rustc verifies that `match` statements are exhaustive. But those ideas are at best half-baked, 
and I'm not seeing an approach that seems reasonable in the short term. So I'm thinking for now we
just have the materialization `source` property be a string for the collection name, unless you see another
approach.

**Comment by jgraettinger** [comment]
Huh, yea, that's a foot-gun. I was thinking mostly of the case where the materialized collection is a union over event types, and you're only interested in events of one kind of variant. But I'm not sure important that really is 🤷 

What about keeping the `source: { name: foo }` structure, but _only_ allowing `name`, and modeling it internally as just source => string. Basically taking your suggestion, but leaving some space in the YAML spec for doing something else later.

**Comment by psFried** [comment]
> What about keeping the source: { name: foo } structure, but only allowing name, and modeling it internally as just source => string. Basically taking your suggestion, but leaving some space in the YAML spec for doing something else later.

I like the idea of representing the `source` the same in both materializations and transforms, since it'll make it more intuitive for people to work with. I was thinking about going the other way and allowing transform `source` to be either a string or an object, so that it could look the same for the common case where you don't have a source schema or partition selectors. But I think I like your idea better, at least for the short term. We do already have a few places in the yaml specs (e.g. `projections`) that allow for strings in a simple case, or objects for more complicated ones. I think it can be a good pattern, but it could also make things more confusing for users if we overuse it. If we start with `source: { name: foo }`, then we could always still allow the plain string representation later if we want. I say all that because I'm a bit on the fence, and could probably be convinced to go with the plain string representation now if you feel like it's worth it.

**Comment by jgraettinger** [comment]
> We do already have a few places in the yaml specs (e.g. projections) that allow for strings in a simple case, or objects for more complicated ones. 

Yea, IMO this is to be avoided. The ergonomics of `projection` make it definitely worthwhile, and `schema` probably so (though even there I have misgivings), but I think `source` definitely doesn't meet the bar, and have a strong preference for requiring `source: { name: foo }` even if that's all it'll ever be.

**Comment by psFried** [comment]
To summarize our conversation from today's VC:

At first blush, I thought it may be desirable to treat some of the items related to error handling
and database identifiers as separate issues. But those changes are required in order to support the
lazy binding that's needed in order to allow collections to be expressed as an object rather than an
array.

Some of the items related to changing the database schema naturally led us to question the use of
sqlite entirely. It feels like the jury is still out on whether it's beneficial to use sqlite to
represent the built catalog, so we decided it's best to keep it for now and just diminish its
primacy so that it is simply an implementation detail.

**Comment by psFried** [comment]
Overall, I'm excited about how this issue is shaping up, and I think catalog specs will be much better off. I had a
few thoughts and questions that I'll list here.

> Allow acyclic imports between catalog resources (implied requirement of derivation cycles).

Not quite sure I understand the meaning here. Does this mean that we want to explicitly disallow
cycles in catalog imports?

> Separate current catalog crate into catalog-models (concerned with representation only), vs catalog-build (concerned with build passes).

I really like this proposed factoring!

Last thing: I'd hate to inflate the scope of this unnecessarily, but I think the changes related to
moving typescript into separate files (as opposed to inlined in the yaml) would also impact the
representation. I think it's ok to ignore those changes for right now, and handle them in a
follow-on issue, but I wanted to bring it up just in case you see a reason to make that part of this
meta-issue.

**Comment by jgraettinger** [comment]
> > Allow acyclic imports between catalog resources (implied requirement of derivation cycles).
> 
> Not quite sure I understand the meaning here. Does this mean that we want to explicitly disallow
> cycles in catalog imports?

Typo changed the meaning (fixed). We currently disallow cycles; they need to be permitted.

**Comment by jgraettinger** [comment]
merged
